packages/belos/tpetra/src/solvers/Belos_Tpetra_GmresSstep.hpp

```diff
@@ -48,32 +47,50 @@ class CholQR {
   FactorOutput
   factor (MV& A, dense_matrix_type& R)
   {
+    Teuchos::RCP< Teuchos::Time > factorTimer = Teuchos::TimeMonitor::getNewCounter ("CholQR::factor");
+    Teuchos::TimeMonitor LocalTimer (*factorTimer);
+
     blas_type blas;
     lapack_type lapack;
 
     const SC zero = STS::zero ();
     const SC one  = STS::one ();
 
-    LO ncols = A.getNumVectors ();
-    LO nrows = A.getLocalLength ();
+    // quick return
+    size_t ncols = A.getNumVectors ();
+    size_t nrows = A.getLocalLength ();
+    if (ncols == 0 || nrows == 0) {
+      return 0;
+    }
 
     // Compute R := A^T * A, using a single BLAS call.
-    MVT::MvTransMv(one, A, A, R);
+    // MV with "static" memory (e.g., Tpetra manages the static GPU memory pool)
+    MV R_mv = makeStaticLocalMultiVector (A, ncols, ncols);
+    //R_mv.putScalar (STS::zero ());
+
+    // compute R := A^T * A
+    R_mv.multiply (Teuchos::CONJ_TRANS, Teuchos::NO_TRANS, one, A, A, zero);
 
     // Compute the Cholesky factorization of R in place, so that
     // A^T * A = R^T * R, where R is ncols by ncols upper
     // triangular.
     int info = 0;
-    lapack.POTRF ('U', ncols, R.values (), R.stride(), &info);
-    if (info < 0) {
-      ncols = -info;
-      // FIXME (mfh 17 Sep 2018) Don't throw; report an error code.
-      throw std::runtime_error("Cholesky factorization failed");
+    R_mv.modify_host ();
+    {
+      auto R_h = R_mv.getLocalViewHost ();
+      lapack.POTRF ('U', ncols, R_h.data (), R_h.extent (0), &info);
```

Comment:
That sounds great. So, we'll be calling cuSolver through Kokkos-kernels?

Reply:
@iyamazaki It looks like kokkos-kernels wants to maintain the cuSOLVER interface, so as soon as they have that interface available, we can use it.  My main worry is that we're switching back and forth between device and host a lot.  If the multiply ends up being on host anyway (because `MPI_Allreduce` doesn't know how to handle CUDA buffers), and if the matrix to factor is small, then it might not be worth the kernel launch overhead to do Cholesky on device.  Thus, I think it would be better to do an experiment first, before we go through the trouble of wrapping cuSOLVER for kokkos-kernels.

Reply:
I am going to take the bait on the cuSOLVER interface comment and respond. I will also include @kddevin as the Tpetra lead. Adding cuSOLVER to Tpetra was incorrect and I said that even when that PR was going in. It was ok to push it because some customer needed it. However, this goes against Trilinos principles of TPL gets interfaced at the right level and in one place. We have violated this basic principle multiple times and that has caused issues with mismatched TPLs. Doing it once for a customer as a **temporary** fix was ok with an assumption it will be fixed. We already have @e10harvey looking at the interfaces to clean that problem on the issue @kddevin filed for Kokkos Kernels. Recommending that pattern to others is just simply not right. 


@iyamazaki did the right thing by requesting TRSM interface and interfacing it right here.

@kddevin Let us talk about this in in the next product leaders meeting on what are Tpetra's plans for design choices with cuSOLVER. I am willing to do cleanup one hot fix, but we cannot do this if using cuSolver is a recommended in PRs.